<a href="https://colab.research.google.com/github/AgustinaStout/Trabajo-Final-Datos/blob/main/Trabajo_final_Agustina_Stout.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Proyecto final Agustina Stout

#Set up
Insalación de las librerias que me van a servir para realizar el trabajo

In [69]:
!pip install pandas==2.2.2

In [70]:
import pandas as pd
import os

#Data Sets
Aca voy a importar los dataset desde mi repositorio de Github. Voy a utilizar head para poder visualizar las primeras cinco filas del Dataset y conocer más sus categorias.

##DataSet Clientes


In [71]:
df_ecommerce_customers_dataset = pd.read_csv("https://raw.githubusercontent.com/AgustinaStout/Trabajo-Final-Datos/main/ecommerce_customers_dataset.csv")
df_ecommerce_customers_dataset.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

##DataSet Ordenes

In [72]:
df_ecommerce_orders_dataset = pd.read_csv("https://raw.githubusercontent.com/AgustinaStout/Trabajo-Final-Datos/main/ecommerce_orders_dataset.csv")
df_ecommerce_orders_dataset.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

##DataSet items


In [73]:
df_ecommerce_order_items_dataset = pd.read_csv("https://raw.githubusercontent.com/AgustinaStout/Trabajo-Final-Datos/main/ecommerce_order_items_dataset.csv")
df_ecommerce_order_items_dataset.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

##DataSet productos


In [74]:
df_ecommerce_products_dataset = pd.read_csv("https://raw.githubusercontent.com/AgustinaStout/Trabajo-Final-Datos/main/ecommerce_products_dataset.csv")
df_ecommerce_products_dataset.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

##Dataset Pagos

In [75]:
df_ecommerce_order_payments_dataset = pd.read_csv("https://raw.githubusercontent.com/AgustinaStout/Trabajo-Final-Datos/main/ecommerce_order_payments_dataset.csv")
df_ecommerce_order_payments_dataset.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

#Analisis básico de los datos
Aca voy a analizar cada uno de los DataSet para obetener datos basicos(DB): Cantidad de filas y columnas (Shape), visualizar las cantidad de datos nulos (isnull().sum()) y qué tipo de datos encontramos (.info()).

In [76]:
def datosDeDataSet(df):
  print("**********")
  print("Tamaño")
  print(df.shape)
  print("**********")
  print("**********")
  print("Cantida de datos nulos")
  print(df.isnull().sum())
  print("**********")
  print("**********")
  print("Tipo de datos")
  print(df.info())
  print("**********")

##DB clientes


In [77]:
datosDeDataSet(df_ecommerce_customers_dataset)

**********
Tamaño
(99441, 5)
**********
**********
Cantida de datos nulos
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
**********
**********
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None
**********


Puedo observar que no hay datos nulos y que casi todos los Data Types son objects.

##DB Ordenes

In [78]:
datosDeDataSet(df_ecommerce_orders_dataset)

**********
Tamaño
(99441, 8)
**********
**********
Cantida de datos nulos
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64
**********
**********
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  

Puedo observar que hay algunos datos nulos y que todos los Data Types son objects.

##DB Items

In [79]:
datosDeDataSet(df_ecommerce_order_items_dataset)

**********
Tamaño
(112650, 7)
**********
**********
Cantida de datos nulos
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64
**********
**********
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB
None
**********


Puedo observar que no hay datos nulos y que casi todos los Data Types son objects.

##DB Productos

In [80]:
datosDeDataSet(df_ecommerce_products_dataset)

**********
Tamaño
(32951, 9)
**********
**********
Cantida de datos nulos
product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64
**********
**********
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   pr

Puedo observar que hay algunos datos nulos y que hay algunos data typo object y otros float.

##DB Pagos

In [81]:
datosDeDataSet(df_ecommerce_order_payments_dataset)


**********
Tamaño
(103886, 5)
**********
**********
Cantida de datos nulos
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64
**********
**********
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB
None
**********


Puedo observar que no hay datos nulos y hay diferentes tipos de Data Types.

#Modificación de datos

En el primer analisis veo que la mayoria de datos aparecen como "object" e "int64". Necesito pasarlos a string y datetime.


##Funciones generales

In [82]:
#Función para cambiar las fechas a Datetime
def transformar_fechas(df, columnas_fechas):
    df[columnas_fechas] = df[columnas_fechas].apply(pd.to_datetime)
    return df

In [83]:
#Función para cambiar los Object que son cadenas a Strings
def transformar_cadenas(df, columnas_cadenas):
  for columna in columnas_cadenas:
    df[columna] = df[columna].astype("string")
  return df

In [84]:
#Función para completar con la media
def completar_media(df, columnas_media):
    for columna in columnas_media:
        df[columna] = df[columna].fillna(df[columna].mean())
    return df

In [85]:
#Función para completar con 0 más de un dato nulo
def fillna_cero(df, columnas):
    df[columnas] = df[columnas].fillna(0)
    return df

##Clientes

In [86]:
#aplico la función para cambiar los datos a String
columnas_cadenas_clientes = ["customer_id","customer_unique_id","customer_city","customer_state"]
df_ecommerce_customers_dataset = transformar_cadenas(df_ecommerce_customers_dataset, columnas_cadenas_clientes)

In [87]:
#Vuelvo a usar la función para revisar los datos
datosDeDataSet(df_ecommerce_customers_dataset)

**********
Tamaño
(99441, 5)
**********
**********
Cantida de datos nulos
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
**********
**********
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  string
 1   customer_unique_id        99441 non-null  string
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  string
 4   customer_state            99441 non-null  string
dtypes: int64(1), string(4)
memory usage: 3.8 MB
None
**********


##Ordenes

In [88]:
#aplico la función para cambiar las fechas a Datetime para las columnas del Dataset Ordenes
columnas_fechas_ordenes = ["order_purchase_timestamp","order_approved_at","order_delivered_carrier_date","order_delivered_customer_date","order_estimated_delivery_date"]
df_ecommerce_orders_dataset = transformar_fechas(df_ecommerce_orders_dataset,columnas_fechas_ordenes)

In [89]:
#Aplico la función para cambiar los datos a strings
columnas_cadenas_ordenes=["order_id","customer_id","order_status"]
df_ecommerce_orders_dataset = transformar_cadenas(df_ecommerce_orders_dataset,columnas_cadenas_ordenes)

In [90]:
#calculo el promedio de diferencia entre columnas de orden de compra y compra arpobada para poder completar los 160 datos nulos
diferencia_tiempo = (df_ecommerce_orders_dataset["order_approved_at"] - df_ecommerce_orders_dataset["order_purchase_timestamp"]).mean()

In [91]:
#completo los nulos con el promedio de las fechas
df_ecommerce_orders_dataset["order_approved_at"]=df_ecommerce_orders_dataset["order_approved_at"].fillna(df_ecommerce_orders_dataset["order_purchase_timestamp"] + diferencia_tiempo)

Para poder rellenar los otros datos nulos que me quedan voy a definir una funcion que me ayude a sacar el promedio de dias que hay entre el momento de la compra y el momento busqueda y entrega. Para luego sumar esa cantidad de dias promedio a las fechas que tengo en la columna de "order_purchase_timestamp" y asi completar los datos nulos de las columnas "order_delivered_carrier_date" y "order_delivered_customer_date".

In [92]:
def calcular_promedio_envio(df, columna_compra, columna_envio):
    tiempo_promedio_envio = (df[columna_envio] - df[columna_compra]).dt.days.mean()

    return tiempo_promedio_envio

In [93]:
#Uso la función para calcular el tiempo promedio de busqueda del envio
promedio_envio = calcular_promedio_envio(df_ecommerce_orders_dataset, "order_purchase_timestamp", "order_delivered_carrier_date")
print("El tiempo promedio de envío es:", promedio_envio, "días")

El tiempo promedio de envío es: 2.7431137234020766 días


In [94]:
df_ecommerce_orders_dataset["order_delivered_carrier_date"] = df_ecommerce_orders_dataset["order_delivered_carrier_date"].fillna(df_ecommerce_orders_dataset["order_purchase_timestamp"] + pd.Timedelta(days=3))

In [95]:
#Uso la función para calcular el tiempo promedio de entrega del envio
promedio_envio = calcular_promedio_envio(df_ecommerce_orders_dataset, "order_purchase_timestamp", "order_delivered_customer_date")
print("El tiempo promedio de entrega es:", promedio_envio, "días")

El tiempo promedio de entrega es: 12.094085575687217 días


In [96]:
df_ecommerce_orders_dataset["order_delivered_customer_date"] = df_ecommerce_orders_dataset["order_delivered_customer_date"].fillna(df_ecommerce_orders_dataset["order_purchase_timestamp"] + pd.Timedelta(days=12))

In [97]:
#Vuelvo a usar la función para revisar los datos
datosDeDataSet(df_ecommerce_orders_dataset)

**********
Tamaño
(99441, 8)
**********
**********
Cantida de datos nulos
order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64
**********
**********
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  string        
 1   customer_id                    99441 non-null  string        
 2   order_status                   99441 non-null  string        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99441 non-null  datetime64[ns]
 5   order_delivered

##Items

In [98]:
#uso la función para cambiar los datos a strings/cadenas
columnas_cadena_items = ["order_id","product_id","seller_id"]
df_ecommerce_order_items_dataset = transformar_cadenas(df_ecommerce_order_items_dataset, columnas_cadena_items)

In [99]:
#uso la funcion para pasar el dato a DateTime
df_ecommerce_order_items_dataset = transformar_fechas(df_ecommerce_order_items_dataset,"shipping_limit_date")

In [100]:
#Vuelvo a usar la función para revisar los datos
datosDeDataSet(df_ecommerce_order_items_dataset)

**********
Tamaño
(112650, 7)
**********
**********
Cantida de datos nulos
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64
**********
**********
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  string        
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  string        
 3   seller_id            112650 non-null  string        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), string(3)
memory usa

##Productos

In [101]:
#uso la función para cambiar los datos a strings/cadenas
columnas_cadenas_productos = ["product_id","product_category_name"]
df_ecommerce_products_dataset = transformar_cadenas(df_ecommerce_products_dataset, columnas_cadenas_productos)

In [102]:
#completo los nulos de la categorias "poduct_category_name" con "Producto desconocido"
df_ecommerce_products_dataset['product_category_name'] = df_ecommerce_products_dataset['product_category_name'].fillna('Producto desconocido')

In [103]:
#aplico funcion para completar con la media los nulos de tres columnas
columnas_a_completar = ['product_name_lenght', 'product_description_lenght', 'product_photos_qty']
df_ecommerce_products_dataset = completar_media(df_ecommerce_products_dataset, columnas_a_completar)

In [104]:
#aplico la función para completar las 4 columnas que tenian pocos nulos con 0, no queria borrarlas
columnas_pesos = ["product_weight_g","product_length_cm","product_height_cm","product_width_cm"]
df_ecommerce_products_dataset = fillna_cero(df_ecommerce_products_dataset, columnas_pesos)

In [105]:
#utilizo nuevamente la función para revisar los datos
datosDeDataSet(df_ecommerce_products_dataset)

**********
Tamaño
(32951, 9)
**********
**********
Cantida de datos nulos
product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64
**********
**********
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  string 
 1   product_category_name       32951 non-null  string 
 2   product_name_lenght         32951 non-null  float64
 3   product_description_lenght  32951 non-null  float64
 4   product_photos_qty          32951 non-null  float64
 5   product_weight_g            32951 non-null  float64
 6   product_length_cm   

##Pagos


In [106]:
#uso la función para cambiar los datos a strings/cadenas
columnas_cadenas_pagos = ["order_id","payment_type"]
df_ecommerce_order_payments_dataset = transformar_cadenas(df_ecommerce_order_payments_dataset, columnas_cadenas_pagos)

In [107]:
#utilizo nuevamente la función para revisar los datos
datosDeDataSet(df_ecommerce_order_payments_dataset)

**********
Tamaño
(103886, 5)
**********
**********
Cantida de datos nulos
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64
**********
**********
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  string 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  string 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), string(2)
memory usage: 4.0 MB
None
**********


#Analisis de DataFrame

In [40]:
#Voy a definir una función que me devuelva datos relevantes en base al dtype que encuentre
def analisis_data(df):
    print("*******")
    print("Análisis variables cuantitativas")
    columnas_numericas = df.select_dtypes(include=['int', 'float']).columns
    if not columnas_numericas.empty:
        print(df[columnas_numericas].describe())
    else:
        print("No hay columnas numéricas en el DataFrame.")

    print("*******")
    print("Análisis variables cualitativas")
    columnas_strings = df.select_dtypes(include=['string']).columns
    if not columnas_strings.empty:
        print(df[columnas_strings].describe())
    else:
        print("No hay columnas cualitativas en el DataFrame.")

    print("*******")
    print("Análisis variables de fecha y hora")
    columnas_datetime = df.select_dtypes(include=['datetime']).columns
    if not columnas_datetime.empty:
        print(df[columnas_datetime].describe())
    else:
        print("No hay columnas de fecha y hora en el DataFrame.")

##Clientes

In [108]:
#Uso la función para ver la info de este DF
analisis_data(df_ecommerce_customers_dataset)

*******
Análisis variables cuantitativas
       customer_zip_code_prefix
count              99441.000000
mean               35137.474583
std                29797.938996
min                 1003.000000
25%                11347.000000
50%                24416.000000
75%                58900.000000
max                99990.000000
*******
Análisis variables cualitativas
                             customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   

       customer_city customer_state  
count          99441          99441  
unique          4119             27  
top        sao paulo             SP  
freq           15540          41746  
*******
Análisis variables de fecha y hora
No hay 

In [42]:
#Utilizo el .value_conts para contar la cantidad de veces que hubo comprar en cada Ciudad
df_ecommerce_customers_dataset["customer_city"].value_counts()

customer_city
sao paulo            15540
rio de janeiro        6882
belo horizonte        2773
brasilia              2131
curitiba              1521
                     ...  
palmares paulista        1
pires ferreira           1
vicente dutra            1
alagoa nova              1
eugenio de castro        1
Name: count, Length: 4119, dtype: Int64

Podemos ver que las principales ciudades que consumen de este comercio son


*   San Paulo
*   Rio de Janeiro
*   Belo Horizonte



In [43]:
conteo_ciudades = df_ecommerce_customers_dataset["customer_city"].value_counts()
ciudades_una_vez = conteo_ciudades[conteo_ciudades == 1]
cantidad_ciudades_una_vez = len(ciudades_una_vez)
print(cantidad_ciudades_una_vez)

1144


Tambien analicé que hay una cantidad de 1144 ciudades que solo compraron una vez, por lo que podrian considerar achicar su rango de llegada para no desperdiciar recursos.

##Ordenes

In [109]:
#Uso la función para ver la info de este DF
analisis_data(df_ecommerce_orders_dataset)

*******
Análisis variables cuantitativas
No hay columnas numéricas en el DataFrame.
*******
Análisis variables cualitativas
                                order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status  
count         99441  
unique            8  
top       delivered  
freq          96478  
*******
Análisis variables de fecha y hora
            order_purchase_timestamp              order_approved_at  \
count                          99441                          99441   
mean   2017-12-31 08:43:12.776581120  2017-12-31 19:08:21.516067840   
min              2016-09-04 21:15:19            2016-09-15 12:16:38   
25%              2017-09-12 14:46:19            

In [45]:
#Voy a sacar el porcentaje de cada Estado que hay en "order_status"
conteo_status = df_ecommerce_orders_dataset["order_status"].value_counts()
total_observaciones = len(df_ecommerce_orders_dataset)
porcentaje_status = (conteo_status / total_observaciones) * 100
print("Porcentaje de cada estado en la columna 'order_status':")
print(porcentaje_status)

Porcentaje de cada estado en la columna 'order_status':
order_status
delivered      97.020344
shipped         1.113223
canceled        0.628513
unavailable     0.612423
invoiced        0.315765
processing      0.302692
created         0.005028
approved        0.002011
Name: count, dtype: Float64


Observamos que la mayoria de las compras ya fueron entregadas, en un 97%. Y solo el 0,6% fue cancelado.

##Items

In [46]:
#Uso la función para ver la info de este DF
analisis_data(df_ecommerce_order_items_dataset)

*******
Análisis variables cuantitativas
       order_item_id          price  freight_value
count  112650.000000  112650.000000  112650.000000
mean        1.197834     120.653739      19.990320
std         0.705124     183.633928      15.806405
min         1.000000       0.850000       0.000000
25%         1.000000      39.900000      13.080000
50%         1.000000      74.990000      16.260000
75%         1.000000     134.900000      21.150000
max        21.000000    6735.000000     409.680000
*******
Análisis variables cualitativas
                                order_id                        product_id  \
count                             112650                            112650   
unique                             98666                             32951   
top     8272b63d03f5f79c56e9e4120aec44ef  aca2eb7d00ea1a7b8ebd4e68314663af   
freq                                  21                               527   

                               seller_id  
count                     

In [47]:
porcentaje_flete = ( 19.990320 / 120.653739) * 100
print("El costo de flete representa aproximadamente el", round(porcentaje_flete, 2), "% del precio del producto.")

El costo de flete representa aproximadamente el 16.57 % del precio del producto.


##Productos


In [48]:
#Uso la función para ver la info de este DF
analisis_data(df_ecommerce_products_dataset)

*******
Análisis variables cuantitativas
       product_name_lenght  product_description_lenght  product_photos_qty  \
count         32951.000000                32951.000000        32951.000000   
mean             48.476949                  771.495285            2.188986   
std              10.150458                  629.208849            1.720614   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  344.000000            1.000000   
50%              51.000000                  604.000000            1.000000   
75%              57.000000                  961.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32951.000000       32951.000000       32951.000000   
mean        2276.334315          30.813207          16.936633   
std         4281.945502          16.915648          13.637779

In [49]:
df_ecommerce_products_dataset["product_category_name"].value_counts().tail(30)

product_category_name
construcao_ferramentas_iluminacao                78
agro_industria_e_comercio                        74
industria_comercio_e_negocios                    68
artigos_de_natal                                 65
audio                                            58
artes                                            55
fashion_underwear_e_moda_praia                   53
dvds_blu_ray                                     48
moveis_quarto                                    45
construcao_ferramentas_ferramentas               39
portateis_casa_forno_e_cafe                      31
livros_importados                                31
pcs                                              30
cine_foto                                        28
musica                                           27
fashion_roupa_feminina                           27
artigos_de_festas                                26
fashion_esporte                                  19
artes_e_artesanato                        

De los últimos 30 productos podemos observar que en su mayoria son de tecnologia, tal vez deberian hacer una revision en profundida con respecto a disminuir la cantidad de productos que se ofrece de esta categoria para poder enforcarse en subir las ventas de las categorias más vendidas.

##Pagos

In [50]:
#Uso la función para ver la info de este DF
analisis_data(df_ecommerce_order_payments_dataset)

*******
Análisis variables cuantitativas
       payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000
*******
Análisis variables cualitativas
                                order_id payment_type
count                             103886       103886
unique                             99440            5
top     fa65dad1b0e818e3ccc5cb0e39231352  credit_card
freq                                  29        76795
*******
Análisis variables de fecha y hora
No hay columnas de fecha y hora en el D

In [51]:
df_ecommerce_order_payments_dataset["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: Int64

El metodo más útilizado para pagar es con tarjeta de credito. Podrian sugerir ofrecer algún beneficio para pagar con tarjeta asi atarae a los otros clientes que utilizan metodos diferentes.

#Nuevos Dataframe


In [52]:
# Crear la carpeta 'dataset_nuevos'
if not os.path.exists('dataset_nuevos'):
    os.makedirs('dataset_nuevos')

##Ordenes-clientes

In [53]:
#voy a hacer un inner merge para poder unir la informacion de los clientes y las ordenes, asi puedo ver mejor el estado de envio, el cliente, cuando y a dónde fue el envio
df_orders_clientes_merge = pd.merge(df_ecommerce_orders_dataset,df_ecommerce_customers_dataset , how = "inner", on= "customer_id")
df_orders_clientes_merge.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                customer_unique_id  \
0                    2017-10-18  7c396fd4830fd04220f754e42b4e5bff   
1                    2018-08-13  af07308b275d755c9edb36a90c618231   
2                    2018-09-04  3a653a41f6f9fc3d2a113cf8398680e8   
3                    2017-12-15  7c142cf63193a1473d2e66489a9ae977   
4                    2018-02-26  72632f0f9dd73dfee390c9b22eb56dd6   

   customer_zip_code_prefix            customer_city customer_state  
0                      3149                sao paulo             SP  
1                     47813                barreiras             BA  
2                     75265               vianopolis             GO  
3                     59296  sao goncalo do amarante             RN  
4                      9195              santo andre             SP

In [54]:
#Para simplificar la información y generar un Dataframe mas claro visualmente y con los criterios que quiero analizar
df_orders_clientes_merge.drop(["order_purchase_timestamp","order_approved_at","order_delivered_carrier_date","order_estimated_delivery_date", "customer_unique_id","customer_zip_code_prefix"], axis=1, inplace=True)
df_orders_clientes_merge.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_delivered_customer_date            customer_city  \
0    delivered           2017-10-10 21:25:13                sao paulo   
1    delivered           2018-08-07 15:27:45                barreiras   
2    delivered           2018-08-17 18:06:29               vianopolis   
3    delivered           2017-12-02 00:28:42  sao goncalo do amarante   
4    delivered           2018-02-16 18:17:02              santo andre   

  customer_state  
0             SP  
1             BA  
2             GO  
3             RN  
4             SP

In [55]:
#por último exporto mi nuevo Dataset como CSV con la función que definí arriba de esta sección.
df_orders_clientes_merge.to_csv("dataset_nuevos/orders_clientes.csv")

##Ordenes-pagos


In [56]:
#voy a hacer un inner merge para poder unir la informacion de las ordenes y pagos, asi puedo ver mejor el estado de envio, el cliente, cuando llego, la forma de pago y lo que se gasto
df_merge_order_items_products = pd.merge(df_ecommerce_orders_dataset, df_ecommerce_order_payments_dataset, how = "inner", on= "order_id")
df_merge_order_items_products.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  payment_sequential payment_type  \
0                    2017-10-18                   1  credit_card   
1                    2017-10-18                   3      voucher   
2                    2017-10-18                   2      voucher   
3                    2018-08-13                   1       boleto   
4                    2018-09-04                   1  credit_card   

   payment_installments  payment_value  
0                     1          18.12  
1                     1           2.00  
2                     1          18.59  
3                     1         141.46  
4                     3         179.12

In [57]:
#Para simplificar la información y generar un Dataframe mas claro visualmente y con los criterios que quiero analizar
df_merge_order_items_products.drop(["order_purchase_timestamp","order_approved_at","order_delivered_carrier_date","order_estimated_delivery_date"], axis=1, inplace=True) #Indicamos el nombre de las columnas y que vamos a mantener los cambios
df_merge_order_items_products.head() #Vemos el resultado

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_delivered_customer_date  payment_sequential payment_type  \
0    delivered           2017-10-10 21:25:13                   1  credit_card   
1    delivered           2017-10-10 21:25:13                   3      voucher   
2    delivered           2017-10-10 21:25:13                   2      voucher   
3    delivered           2018-08-07 15:27:45                   1       boleto   
4    delivered           2018-08-17 18:06:29                   1  credit_card   

   payment_installments  payment_value  
0                     1          18.12  
1                     1           2.00  
2                     1          18.59  
3                     1         141.46  
4                     3         179.12

In [58]:
#por último exporto mi nuevo Dataset como CSV con la función que definí arriba de esta sección.
df_merge_order_items_products.to_csv("dataset_nuevos/orders_items.csv")

##Items-productos

In [59]:
#voy a hacer un inner merge para poder unir la informacion de los items y productos, asi puedo ver mejor el producto,su precio 7 su peso
df_productos_items_merge = pd.merge(df_ecommerce_order_items_dataset, df_ecommerce_products_dataset, how="inner", on= "product_id")
df_productos_items_merge.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

  shipping_limit_date   price  freight_value product_category_name  \
0 2017-09-19 09:45:35   58.90          13.29            cool_stuff   
1 2017-05-03 11:05:13  239.90          19.93              pet_shop   
2 2018-01-18 14:48:30  199.00          17.87      moveis_decoracao   
3 2018-08-15 10:10:18   12.99          12.79            perfumaria   
4 2017-02-13 13:57:51  199.90          18.14    ferramentas_jardim   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 58.0                       598.0                 4.0   
1                 56.0                       239.0                 2.0   
2                 59.0                       695.0                 2.0   
3                 42.0                       480.0                 1.0   
4                 59.0                       409.0                 1.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             650.0               28.0                9.0              14.0  
1           30000.0               50.0               30.0              40.0  
2            3050.0               33.0               13.0              33.0  
3             200.0               16.0               10.0              15.0  
4            3750.0               35.0               40.0              30.0

In [60]:
#Para simplificar la información y generar un Dataframe mas claro visualmente y con los criterios que quiero analizar
df_productos_items_merge.drop(["seller_id","shipping_limit_date","product_name_lenght","product_description_lenght","product_photos_qty","product_length_cm","product_height_cm","product_width_cm"], axis=1, inplace=True)
df_productos_items_merge.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id   price  freight_value  \
0  4244733e06e7ecb4970a6e2683c13e61   58.90          13.29   
1  e5f2d52b802189ee658865ca93d83a8f  239.90          19.93   
2  c777355d18b72b67abbeef9df44fd0fd  199.00          17.87   
3  7634da152a4610f1595efa32f14722fc   12.99          12.79   
4  ac6c3623068f30de03045865e4e10089  199.90          18.14   

  product_category_name  product_weight_g  
0            cool_stuff             650.0  
1              pet_shop           30000.0  
2      moveis_decoracao            3050.0  
3            perfumaria             200.0  
4    ferramentas_jardim            3750.0

In [61]:
#por último exporto mi nuevo Dataset como CSV con la función que definí arriba de esta sección.
df_productos_items_merge.to_csv("dataset_nuevos/items_productos.csv")